torch/nn/parallel/distributed.py

```diff
@@ -130,31 +136,60 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
+            for p in module.parameters():
+                tp = p.type()
+                if tp == torch.cuda.HalfTensor and \
+                        dist._backend != dist.dist_backend.NCCL and \
+                        dist._backend != dist.dist_backend.GLOO:
+                    raise RuntimeError("DistributedDataParallel currently only "
+                                       "supports half precision parameters "
+                                       "with NCCL backend")
+                # Add the parameter into the type bucket
+                self.param_type_buckets[tp][device_idx].append(p)
+
+        # TODO, adding mixed precision support in NCCL reduction code path
+        # This is because NCCL backend doesn't support multiple reduction
+        # bucket
+        if len(self.param_type_buckets) > 1 and \
+                dist._backend == dist.dist_backend.NCCL:
+            raise RuntimeError("DistributedDataParallel currently doesn't "
+                               "support mixed precision type for NCCL backend")
+
+        # Split parameters into buckets that will coalesce reductions
+        #
+        # Note that previously we have already splitted parameters by the type.
+        # Here, for each type, we further split each type of parameters into
+        # reduction buckets so that each bucket will only have a single type
+        # of parameters. Therefore subsequent all-reduce can be successful since
+        # the reduction operation needs to operate on the same kind of data type
         self.bucket_sizes = []
         self.bucket_map = {}
+
         # Currently NCCL backend only supports single reduction thread/bucket
         if dist._backend == dist.dist_backend.NCCL:
             bucket_bytes_cap = float('inf')
         else:
             bucket_bytes_cap = 1 * MB
-        bucket_bytes = bucket_bytes_cap  # to init the first bucket immediately
-        for param_tuple in zip(*map(lambda m: m.parameters(), self._module_copies)):
-            if param_tuple[0].requires_grad:
+
+        for tp in self.param_type_buckets:
+            # to init the first bucket immediately for each type
+            bucket_bytes = bucket_bytes_cap
+            for param_idx, param in enumerate(self.param_type_buckets[tp][0]):
+                if not param.requires_grad:
+                    continue
                 if bucket_bytes >= bucket_bytes_cap:
                     self.bucket_sizes.append(0)
                     bucket_bytes = 0
-                for p in param_tuple:
-                    self.bucket_map[p] = len(self.bucket_sizes) - 1
-                bucket_bytes += p.numel() * p.element_size()
+                for dev_idx in range(len(self.device_ids)):
+                    dev_param = self.param_type_buckets[tp][dev_idx][param_idx]
+                    self.bucket_map[dev_param] = len(self.bucket_sizes) - 1
+                bucket_bytes += param.numel() * param.element_size()
                 self.bucket_sizes[-1] += 1
```

Comment:
Actually can't we just use `list(take_tensors(module.parameters()))` for this? This would remove a lot of code in here and simplify DDP a lot

Reply:
We still need to get the size etc, I think it's better to keep the original logic. This new logic is  also tested

Reply:
I disagree. It's always better to simplify the code, especially that we already have good tools for this. Getting the size isn't a problem, just take the `len` of the thing that `_take_tensors` gave you.

Reply:
Ok, I changed the entire code, but found that _take_tensors doesn't work with Variables, it does bucket Variables by size, but not by types.